saga/sagalogs/memory.go

```diff
@@ -0,0 +1,155 @@
+
+// Returns an Instance of a Saga based on an inMemorySagaLog.
+// This is an in-memory impl, and is not durable.
+// Implements GC of whole sagas based on time expiration.
+// GC should be set at a duration that realistically will not purge active Sagas.
+// gcExpiration: duration after which a Saga was created, it will be deleted.
+// 	A zero duration is interpretted as "never gc" (the Log will eventually consume all memory).
+// gcInterval: duration interval at which GC runs.
+func MakeInMemorySagaCoordinator(gcExpiration time.Duration, gcInterval time.Duration) saga.SagaCoordinator {
+	return saga.MakeSagaCoordinator(MakeInMemorySagaLog(gcExpiration, gcInterval))
+}
+
+func MakeInMemorySagaLog(gcExpiration time.Duration, gcInterval time.Duration) saga.SagaLog {
+	slog := &inMemorySagaLog{
+		sagas:        make(map[string]*logData),
+		mutex:        sync.RWMutex{},
+		gcExpiration: gcExpiration,
+		gcTicker:     time.NewTicker(gcInterval),
+	}
+	go func() {
+		if gcExpiration != 0 {
+			for range slog.gcTicker.C {
+				if err := slog.gcSagas(); err != nil {
+					log.Errorf("Error running gcSagas: %s", err)
+				}
+			}
+		}
+	}()
+	return slog
+}
+
+// Creates a Saga in the log and adds a StartSagaMessage to it
+func (slog *inMemorySagaLog) StartSaga(sagaId string, job []byte) error {
+	slog.mutex.Lock()
+	defer slog.mutex.Unlock()
+
+	startMsg := saga.MakeStartSagaMessage(sagaId, job)
+	slog.sagas[sagaId] = &logData{messages: []saga.SagaMessage{startMsg}, created: time.Now()}
+
+	return nil
+}
+
+// Log a SagaMessage to an existing Saga in the log
+func (slog *inMemorySagaLog) LogMessage(msg saga.SagaMessage) error {
+	slog.mutex.Lock()
+	defer slog.mutex.Unlock()
+
+	sagaId := msg.SagaId
+	ld, ok := slog.sagas[sagaId]
+	if !ok {
+		return errors.New(fmt.Sprintf("Saga: %s does not exist in the Log", sagaId))
+	}
+
+	ld.messages = append(ld.messages, msg)
+	return nil
+}
+
+// Gets all SagaMessages from an existing Saga in the log
+func (slog *inMemorySagaLog) GetMessages(sagaId string) ([]saga.SagaMessage, error) {
+	slog.mutex.RLock()
+	defer slog.mutex.RUnlock()
+
+	ld, ok := slog.sagas[sagaId]
+	if ok {
+		return ld.messages, nil
+	} else {
+		return nil, nil
+	}
+}
+
+// Returns all non-GCd SagaIds existing since this SagaLog was created.
+// Includes Sagas of any state (completed, active, etc).
+func (slog *inMemorySagaLog) GetActiveSagas() ([]string, error) {
+	slog.mutex.RLock()
+	defer slog.mutex.RUnlock()
+
+	keys := make([]string, len(slog.sagas))
+	i := 0
+
+	for key := range slog.sagas {
+		keys[i] = key
+		i++
+	}
+
+	return keys, nil
+}
+
+// Check for expired Sagas and then delete them
+func (slog *inMemorySagaLog) gcSagas() error {
+	expired := slog.getExpiredSagaIds()
+	if len(expired) == 0 {
+		return nil
+	}
+
+	slog.mutex.Lock()
+	defer slog.mutex.Unlock()
+
+	for _, id := range expired {
+		delete(slog.sagas, id)
+	}
+
+	return nil
+}
+
+func (slog *inMemorySagaLog) getExpiredSagaIds() []string {
+	slog.mutex.RLock()
```

Comment:
should we only gc completed saga's (ones with endSaga)?

Reply:
I was on the fence here. I initially thought about having a hard and soft cap - clean up completed sagas after a soft interval and any after a hard interval. I think in practice, though, we don't have a use case for that, and decided to keep it simpler. Do you have a strong preference?

Reply:
as long as we clearly document that gc collects any saga (regardless of state), then as is is fiine.

Reply:
added more clear comments